Update location through user api

jingcheng16 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
This PR add logic to update location for mobile workers through user api. 

- get mobile worker through api will also return `primary_location` and `locations`
- create new mobile worker and update existing mobile worker now accept two new fields: `primary_location` and `locations`.
- if the two fields is not included in the json object, nothing will happen.
- if the two fields is included, `primary_location` is set to empty string, and `locations` is set to empty array, then all locations and primary location will be removed from that user.

Criteria on updating locations:
- the two new fields must be provided together
- `primary_location` must be one of `locations`
- all location must be valid

Updated wiki: (Please don't click on Update)
List Mobile Workers API: https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/edit-v2/2143957296
User Creation: https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/edit-v2/2143957350
User Edit: https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/edit-v2/2143957433

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
Ticket: https://dimagi.atlassian.net/browse/SAAS-15722
This PR introduces changes to the CommCare User API, mimicking the UI of change location for a mobile worker: it will always take two fields, `primary_location` (a string that represents location id of the primary location) and `locations` (an array of location ids) together.
The update logic ensures that both fields must be provided together, with validation ensuring that the primary location is included in the list of locations, all location are valid. 
Then I will take the two field to one location_obj dict, and pass it to `update` function, which will be called in both create mobile worker and edit mobile worker functions.


## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
I tested the changes locally and on staging. 

Validation has been added to prevent incorrect or incomplete updates.

I think I maintained backward compatibility, if this two field is not included, then the api will still work just as how it works before without editing locations of mobile workers. But if user is using the returned data of GET request for further process, I did return two more fields: `locations` and `primary_location`, will this be a concern?

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
corehq/apps/api/tests/test_user_resources.py
corehq/apps/api/tests/test_user_updates.py

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
QA Ticket: https://dimagi.atlassian.net/browse/QA-7085

QA Plan:
- Test List Mobile Workers api, and verify, `primary_location` and `locations` are now included in the returned results
- Test Create Mobile Worker api and verify mobile workers with locations can be created
- Test Edit Mobile Worker api and verify mobile workers's location can be updated
      - When no location field included, then locations won't be updated
      - When only `primary_location` or `locations` included, then error is raised
      - When invalid location passed, then error is raised
      - When `primary_location` is not in the list of `locations`, error is raised
- Prevent Location Updates in user_data:
Ensure that location information cannot be overwritten by putting `commcare_location_ids` and `commcare_location_id` in user_data. The update will be rejected and both the location and user_data remains unchanged. 
-  A regression test on other user apis

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

jingcheng16 committed:
Rename user_resources to test_user_resources

jingcheng16 committed:
Main logic to enable location update

jingcheng16 committed:
Add test for updating location through user api

esoergel commented on corehq/apps/api/user_updates.py:
```diff
@@ -165,3 +167,53 @@ def _log_phone_number_change(new_phone_numbers, old_phone_numbers, user_change_l
+            user_change_logger.add_changes({'assigned_location_ids': location_ids})
+            user_change_logger.add_info(UserChangeMessage.assigned_locations_info([]))
+    elif not primary_location_id or not location_ids:
+        raise UpdateUserException(_('Both primary_location and locations must be provided together.'))
```
I know we discussed this kind of thing a bunch - this approach seems appropriate to me, good to default to the more explicit option, especially in an API

esoergel commented on corehq/apps/api/user_updates.py:
```diff
@@ -165,3 +167,53 @@ def _log_phone_number_change(new_phone_numbers, old_phone_numbers, user_change_l
         user_change_logger.add_change_message({'phone_numbers': change_messages})
+
+
+def _update_location(user, location_object, user_change_logger):
```
What would you think about pulling parts of this out into a couple helper functions?

jingcheng16 replied:
refactored in ca474379fe01a01e967cdc77dbf5d8d24ad1c64d

esoergel commented on corehq/apps/api/user_updates.py:
```diff
@@ -165,3 +167,53 @@ def _log_phone_number_change(new_phone_numbers, old_phone_numbers, user_change_l
+        if primary_location_id not in location_ids:
+            raise UpdateUserException(_('Primary location must be included in the list of locations.'))
+
+        # Verify all location ids has a SQLLocation.
+        for loc_id in location_ids:
+            try:
+                SQLLocation.active_objects.get(location_id=loc_id, domain=user.domain)
+            except SQLLocation.DoesNotExist:
+                raise UpdateUserException(f"Could not find location id '{loc_id}'.")
```
I believe you could also do this in bulk with something like this:
```suggestion
        real_ids = (SQLLocation.active_objects
                    .filter(location_id__in=location_ids, domain=user.domain)
                    .location_ids())
        if missing_ids:= set(location_ids) - set(real_ids):
            raise UpdateUserException(f"Could not find location ids {', '.join(missing_ids)}.")
```

jingcheng16 replied:
Thank you! Did in af1bb910b2789f0432a3df36a80e6ff28c9a4a2d

esoergel commented on corehq/apps/api/user_updates.py:
```diff
@@ -165,3 +167,53 @@ def _log_phone_number_change(new_phone_numbers, old_phone_numbers, user_change_l
+                raise UpdateUserException(f"Could not find location id '{loc_id}'.")
+
+        if primary_location_id != user_current_primary_location_id:
+            primary_location = SQLLocation.active_objects.get(location_id=primary_location_id)
+            user.set_location(primary_location, commit=False)
```
`set_location` also takes a location ID, so you can skip the lookup.  Or is it necessary for the logging?  If you do need the location objects anyways, then I'd just fetch them all a couple lines above when you check that they're real, to avoid extra DB hits.

jingcheng16 replied:
It is necessary for the logging. I just realized I query all the locations twice, so I updated in 106745fccd65bba25d3a47c9be960b01ae8e362c. But I don't know how to get the `primary_location` object from that query? The only thing I can think of is iterate through all the location objects, then compare the location id and `primary_location_id`

esoergel reviewed (commented):
Really excited to see this coming through

jingcheng16 committed:
Verify location ids in bulk

jingcheng16 committed:
less db hits

jingcheng16 committed:
REFACTOR: break _update_locations into smaller functions

jingcheng16 committed:
Add test test update should not run if there is no change

gherceg commented on corehq/apps/api/user_updates.py:
```diff
@@ -194,9 +194,9 @@ def _update_location(user, location_object, user_change_logger):
-                    .filter(location_id__in=location_ids, domain=user.domain)
-                    .location_ids())
+        locations = SQLLocation.active_objects.filter(location_id__in=location_ids, domain=user.domain)
+        real_ids = locations.location_ids()
```
I think this is effectively the same as before since Django [querysets are lazy](https://docs.djangoproject.com/en/4.2/topics/db/queries/#querysets-are-lazy), and this doesn't take advantage of [QuerySet caching](https://docs.djangoproject.com/en/4.2/topics/db/queries/#caching-and-querysets).

You want to force the evaluation of the base query, in this case `locations = SQLLocation.active_objects.filter(location_id__in=location_ids, domain=user.domain)`
```suggestion
        locations = SQLLocation.active_objects.filter(location_id__in=location_ids, domain=user.domain)
        _ = list(locations)  # forces evaluation and populates locations._result_cache which can be used by the next line
        real_ids = locations.location_ids()
```

There might be a cleaner way to force evaluation of the queryset, but you want to keep `locations` as a QuerySet so that it can be used in both the context of fetching IDs specifically, and then later on to get the objects (which you would have already done when forcing the evaluation up front).

jingcheng16 replied:
Thank you. Updated in 23bbc659ec2c18189208150dc3599361dd95abef

gherceg commented on corehq/apps/api/user_updates.py:
```diff
@@ -176,42 +176,60 @@ def _update_location(user, location_object, user_change_logger):
         return
 
-    user_current_primary_location_id = user.location_id
+    user_current_primary_location_id = user.get_location_id(user.domain)
+    user_current_locations = user.get_location_ids(user.domain)
```
nit: I think you can remove `user_` from the front of these names as we are operating in the context of updating locations on a specific user so it is already clear.

jingcheng16 replied:
Removed in 30c498d7481b28020502f54341ff09c4866ffe20

gherceg reviewed (commented):
Only looked over the newer commits.

jingcheng16 committed:
Force evaluation of locations queryset

jingcheng16 committed:
REFACTOR: renaming variable

remove user_ from the front of these names as we are operating in the context of updating locations on a specific user so it is already clear.

gherceg commented on corehq/apps/api/user_updates.py:
```diff
@@ -165,3 +167,70 @@ def _log_phone_number_change(new_phone_numbers, old_phone_numbers, user_change_l
+
+def _verify_location_ids(location_ids, domain):
+    locations = SQLLocation.active_objects.filter(location_id__in=location_ids, domain=domain)
+    _ = list(locations)  # forces evaluation and populates locations._result_cache which can be used later
+    real_ids = locations.location_ids()
```
```suggestion
    real_ids = [loc.location_id for loc in locations]
```

Ok turns out I was still wrong and `locations.values_list(...)` will clone the query but not the result cache, therefore not using the cache. (`location_ids` calls `values_list`). So let's just do this in python instead.

jingcheng16 committed:
Refactoring: use the cache in locations

Co-authored-by: Graham Herceg <gherceg@dimagi.com>

gherceg reviewed (approved):
Just nitpicks

jingcheng16 committed:
Refactor: fixing test name

Co-authored-by: Graham Herceg <gherceg@dimagi.com>
